### PR TITLE
ixx: swap backend to ixx

### LINKED ISSUE #24: Markdown bullet list not rendered
ours: https://nuschtos-search.pages.dev/?query=nixvim&option=programs.nixvim.autoCmd.*.callback
original: https://nix-community.github.io/nixvim/NeovimOptions/autoCmd/index.html?highlight=autoCmd.*.callback#autocmdcallback

### LINKED ISSUE #84: Add option to filter by scope

### LINKED ISSUE #105: improve index
Improve the index to support massive deployments. (the goal is to be able to search the whole of nixpkgs on the frontend)

- Separate index file in binary format
- Compressed like labels in DNS
- webasably/rust based search engine -> search directly on the compressed data

### PR DESCRIPTION
**Pull request by MarcelCoding**
closes #24
closes #84
closes #105

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,34 +12,35 @@
...
+    "@angular/language-service": "^18.2.8",
     "@types/jasmine": "~5.1.4",
     "jasmine-core": "~5.3.0",
     "karma": "~6.4.4",
     "karma-chrome-launcher": "~3.2.0",
     "karma-coverage": "~2.2.1",
     "karma-jasmine": "~5.1.0",
     "karma-jasmine-html-reporter": "~2.1.0",
-    "typescript": "~5.5.4"
+    "typescript": "~5.4.5"
```
Comment (SuperSandro2000) [inline package.json:44]: Why the downgrade?

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -1,107 +1,118 @@
-lockfileVersion: '9.0'
+lockfileVersion: '6.0'
```
Comment (SuperSandro2000) [inline pnpm-lock.yaml:1]: Lets keep pnpm_9

File: src/app/core/components/search/search.component.ts [hunk 3]
```diff
@@ -25,13 +25,10 @@ export class SearchComponent implements AfterViewInit {
     switchMap(query => {
       const q = query?.trim();
       if (q !== getQuery()) this.router.navigate([], { queryParams: { query: q }, queryParamsHandling: 'merge' });
-      if (q && q.length > 0) {
         return this.searchService.search(q);
```
Comment (SuperSandro2000) [inline src/app/core/components/search/search.component.ts:28]:
```suggestion
      return this.searchService.search(q);
```

File: src/app/core/data/search.service.ts [hunk 4]
```diff
@@ -18,79 +26,58 @@ export interface Option {
...
-    }
-
-    return this.data.pipe(map(options => {
-      const result = [];
+  public search(query: string | null | undefined): Observable<SearchedOption[]> {
+    return this.index.pipe(
+      map(index => {
+        return index ? (query && query.length > 0 ? index.search(query, 500).map(option => {
+          const opt = ({ idx: option.idx(), name: option.name() });
+          //      option.free();
```
Comment (SuperSandro2000) [inline src/app/core/data/search.service.ts:45]: C + typescritp + rust 😂

File: src/app/core/data/search.service.ts [hunk 5]
```diff
@@ -18,79 +26,58 @@ export interface Option {
...
-    if (search.length === 0) {
-      return of([]);
-    }
-
-    return this.data.pipe(map(options => {
-      const result = [];
+  public search(query: string | null | undefined): Observable<SearchedOption[]> {
+    return this.index.pipe(
+      map(index => {
+        return index ? (query && query.length > 0 ? index.search(query, 500).map(option => {
```
Comment (SuperSandro2000) [inline src/app/core/data/search.service.ts:43]: We should move the 500 into a const, too.

File: package.json (hunk 1 above)
Comment (MarcelCoding) [reply package.json:44]: Probably just merge conflicts

File: src/app/core/data/search.service.ts (hunk 4 above)
Comment (MarcelCoding) [reply src/app/core/data/search.service.ts:45]: According to the documentation free should be called when the strict is not needed anymore but calling free results in some pointer error... Need to look into this

File: src/app/core/data/search.service.ts (hunk 5 above)
Comment (MarcelCoding) [reply src/app/core/data/search.service.ts:43]: Yeah, I think it is currently referenced twice.

File: flake.nix [hunk 6]
```diff
@@ -4,28 +4,37 @@
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
     flake-utils.url = "github:numtide/flake-utils";
+    ixx = {
+      # match version with npm package
+      url = "github:/NuschtOS/ixx/declaration-store-path";
```
Comment (SuperSandro2000) [inline flake.nix:9]:
```suggestion
      url = "github:NuschtOS/ixx/declaration-store-path";
```

File: nix/fixup-options.py
[the remark this answers is not in the document]
Comment (MarcelCoding) [reply nix/fixup-options.py:1]: yeah :tada:

File: flake.nix [hunk 7]
```diff
@@ -4,28 +4,37 @@
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
     flake-utils.url = "github:numtide/flake-utils";
+    ixx = {
+      # match version with npm package
+      url = "github:NuschtOS/ixx/v0.0.4";
```
Comment (SuperSandro2000) [inline flake.nix:9]: We should somehow assert this or add an update script or so

File: src/app/core/components/search/search.component.html [hunk 8]
```diff
@@ -9,5 +19,6 @@
       </a>
     </li>
   </ul>
-  <p *ngIf="options.length == 500">Only showing the first 500 results. Make your search term more concise.</p>
+  <p *ngIf="options.length == maxSearchResults">Only showing the first {{maxSearchResults}} results. Make your search
+    term more concise.</p>
```
Comment (SuperSandro2000) [inline src/app/core/components/search/search.component.html:23]:
```suggestion
  <p *ngIf="options.length == maxSearchResults">Only showing the first {{maxSearchResults}} results. Make your search term more concise.</p>
```

Comment (MarcelCoding) [reply src/app/core/components/search/search.component.html:23]: Why remove the line break, the formatter will put it back in?

File: flake.nix (hunk 7 above)
Comment (MarcelCoding) [reply flake.nix:9]: Can this be done without embedding the version in the flake itself? Pasing the url? In development or should be fine to not use the exact same version .

Comment (SuperSandro2000) [reply flake.nix:9]: We cannot add more fanciness to the URL as that part of nix code is really restricted.

We probably need to add some assertion to the nix code which builds the package or so. or another update script.

File: src/app/core/components/search/search.component.html (hunk 8 above)
Comment (SuperSandro2000) [reply src/app/core/components/search/search.component.html:23]: We have 80 line length? ...

Comment (MarcelCoding) [reply src/app/core/components/search/search.component.html:23]: idk its a thing of html formatters

**Comment by SuperSandro2000** [comment]
We should not use a number to track the scopes otherwise the URLs are not stable

**Comment by MarcelCoding** [comment]
Any recommendations I would not like to introduce a separate but the url and name are not stable itself

**Comment by SuperSandro2000** [comment]
The name is probably as stable as we get it. The whole url is to long and adding an extra option is probably overkill. 

Maybe we should add a fallback if the selected option is no longer in the scope. Something like did you mean XYZ.

**Pull request merged**